File: cadwiki-nuget/cadwiki.AutoCAD2021.Base.Utilities/Shared/Paperspace.vb
Option Explicit On
Option Strict On
Option Infer Off

Imports Autodesk.AutoCAD.ApplicationServices
Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.Geometry
Imports Autodesk.AutoCAD.GraphicsSystem

Public Class Paperspace
    Public Shared Function CreateViewPort(doc As Document) As ObjectId
        Dim db As Database = doc.Database
        Using lock As DocumentLock = doc.LockDocument
            Using t As Transaction = db.TransactionManager.StartTransaction()
                Dim bt As BlockTable = CType(t.GetObject(db.BlockTableId, OpenMode.ForRead), BlockTable)
                Dim rec As BlockTableRecord = CType(t.GetObject(bt(BlockTableRecord.PaperSpace), OpenMode.ForWrite),
                    BlockTableRecord)

                Application.SetSystemVariable("TILEMODE", 0)
                doc.Editor.SwitchToPaperSpace()

                Dim vp As Viewport = New Viewport()
                vp.SetDatabaseDefaults()
                vp.CenterPoint = New Point3d(3.25, 3, 0)
                vp.Width = 6
                vp.Height = 5

                Dim objectId As ObjectId = rec.AppendEntity(vp)
                t.AddNewlyCreatedDBObject(vp, True)
                vp.On = True

                t.Commit()
                Return objectId
            End Using
        End Using


        Return Nothing
    End Function


    Private Shared Sub ZoomCurrentViewPort(doc As Document, lowerLeft As Point3d, upperRight As Point3d)
        Dim cvId As Integer = Convert.ToInt32((Application.GetSystemVariable("CVPORT")))
        Using gm As Manager = doc.GraphicsManager
            Using vw As View = gm.GetCurrentAcGsView(cvId)
                Dim currentTab As String = CStr(Application.GetSystemVariable("CTAB"))
                Layouts.SetCurrentTab(doc, "Model")
                vw.ZoomExtents(lowerLeft, upperRight)
                vw.Zoom(0.95)
                'SetCurrentLayoutTab(currentTab)
                'gm.SetViewportFromView(cvId, vw, True, True, False)
            End Using
        End Using
    End Sub
End Class

File: cadwiki-nuget/cadwiki.AC/Utilities/Paperspace.cs
using System;
using Autodesk.AutoCAD.ApplicationServices;
using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.Geometry;
using Microsoft.VisualBasic.CompilerServices;

namespace cadwiki.AC
{

    public class Paperspace
    {
        public static ObjectId CreateViewPort(Document doc)
        {
            var db = doc.Database;
            using (var @lock = doc.LockDocument())
            {
                using (var t = db.TransactionManager.StartTransaction())
                {
                    BlockTable bt = (BlockTable)t.GetObject(db.BlockTableId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForRead);
                    BlockTableRecord rec = (BlockTableRecord)t.GetObject(bt[BlockTableRecord.PaperSpace], global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);

                    global::Autodesk.AutoCAD.ApplicationServices.Core.Application.SetSystemVariable("TILEMODE", 0);
                    doc.Editor.SwitchToPaperSpace();

                    var vp = new Viewport();
                    vp.SetDatabaseDefaults();
                    vp.CenterPoint = new Point3d(3.25d, 3d, 0d);
                    vp.Width = 6d;
                    vp.Height = 5d;

                    var objectId = rec.AppendEntity(vp);
                    t.AddNewlyCreatedDBObject(vp, true);
                    vp.On = true;

                    t.Commit();
                    return objectId;
                }
            }


            return default;
        }


        private static void ZoomCurrentViewPort(Document doc, Point3d lowerLeft, Point3d upperRight)
        {
            int cvId = Convert.ToInt32(global::Autodesk.AutoCAD.ApplicationServices.Core.Application.GetSystemVariable("CVPORT"));
            using (var gm = doc.GraphicsManager)
            {
                using (var vw = gm.GetCurrentAcGsView(cvId))
                {
                    string currentTab = Conversions.ToString(global::Autodesk.AutoCAD.ApplicationServices.Core.Application.GetSystemVariable("CTAB"));
                    Layouts.SetCurrentTab(doc, "Model");
                    vw.ZoomExtents(lowerLeft, upperRight);
                    vw.Zoom(0.95d);
                    // SetCurrentLayoutTab(currentTab)
                    // gm.SetViewportFromView(cvId, vw, True, True, False)
                }
            }
        }
    }
}